### PR TITLE
Migration to ibc_proto v.0.3.0

### PR DESCRIPTION
**Pull request by ancazamfir**
Closes: #226

## Description
Migrate to ibc_proto v0.3.0.

Includes minimal changes such that the code compiles and tests pass. In particular:

- the client state in the handshake messages is not deserialized and therefore further ignored
- the epoch in the new type `Height` is ignored and the deserialization code only extracts the `EpochHeight` in the current `Height` type which is u64.

______

For contributor use:

- [ ] Unit tests written
- [ ] Added test to CI if applicable 
- [ ] Updated CHANGELOG_PENDING.md
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: modules/src/ics03_connection/error.rs
Comment (romac) [inline modules/src/ics03_connection/error.rs:42]:
```suggestion
    MissingConsensusHeight,
```

File: modules/src/ics03_connection/error.rs
Comment (romac) [inline modules/src/ics03_connection/error.rs:41]:
```suggestion
    #[error("missing consensus proof height in openTry")]
```

File: modules/src/ics03_connection/error.rs
Comment (romac) [inline modules/src/ics03_connection/error.rs:38]:
```suggestion
    #[error("missing consensus proof height in openTry")]
```

File: modules/src/ics03_connection/handler/verify.rs
Comment (romac) [inline modules/src/ics03_connection/handler/verify.rs:41]:
```suggestion
    if let Some(state) = client_state {
        verify_client_proof(
            ctx,
            connection_end,
            state,
            proofs.height(),
            proofs
                .client_proof()
                .as_ref()
                .ok_or_else(|| Kind::NullClientProof)?,
        )?;
    }
```

File: modules/src/ics03_connection/msgs.rs
Comment (romac) [inline modules/src/ics03_connection/msgs.rs:240]:
```suggestion
            .ok_or_else(|| Kind::MissingConsensusHeight)?
```

File: modules/src/ics03_connection/msgs.rs
Comment (romac) [inline modules/src/ics03_connection/msgs.rs:360]:
```suggestion
            .ok_or_else(|| Kind::MissingConsensusHeight)?
```

**Review by romac — COMMENTED** [review]
Looks good, modulo some typos and a couple idioms :) Will leave it to @adizere to approve the PR as I am not super familiar with ICS 03 yet.

**Comment by adizere** [comment]
Heads-up, once we merge #220 we might get the following error:

```
$ cargo clippy --all-targets
    Checking ibc v0.0.3 (/Users/adi/Hammers/ibc-rs/modules)
error: large size difference between variants
  --> modules/src/ics03_connection/msgs.rs:51:5
   |
51 |     ConnectionOpenTry(MsgConnectionOpenTry),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this variant is 352 bytes
   |
note: the lint level is defined here
  --> modules/src/lib.rs:2:9
   |
2  | #![deny(clippy::all)]
   |         ^^^^^^^^^^^
   = note: `#[deny(clippy::large_enum_variant)]` implied by `#[deny(clippy::all)]`
note: and the second-largest variant is 144 bytes:
  --> modules/src/ics03_connection/msgs.rs:50:5
   |
50 |     ConnectionOpenInit(MsgConnectionOpenInit),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
help: consider boxing the large fields to reduce the total size of the enum
   |
51 |     ConnectionOpenTry(Box<MsgConnectionOpenTry>),
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: could not compile `ibc`.```

File: modules/src/ics02_client/state.rs
Comment (adizere) [inline modules/src/ics02_client/state.rs:47]: The other two `verify_*` functions here appear in [ICS02](https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics#required-functions) but this `verify_client_full_state ` one does not ring a bell. All implementations of `verify_client_full_state` are also `unimplemented` so it's not really clear what the function does by just looking at the code. Can we have a comment here (describing the purpose of the function like we do for `verify_client_consensus_state` or `verify_connection_state`)?

**Review by adizere — APPROVED** [review]
I think `clippy` will bite us with next commits, so we'll need to address that (see my earlier comment). Otherwise looks good!

File: modules/src/ics03_connection/msgs.rs
Comment (brapse) [inline modules/src/ics03_connection/msgs.rs:747]: Why the indentation here?

**Comment by ancazamfir** [comment]
Another need to fix before merge:

When running the Rust relayer query with Gaia and Go Relayer, the connection query fails. 
```
cargo run --bin relayer -- -c simple_config.toml query connection end ibc0 ibconeconnection
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running `target/debug/relayer -c simple_config.toml query connection end ibc0 ibconeconnection`
     Options QueryConnectionOptions { connection_id: ConnectionId("ibconeconnection"), height: 0, proof: true }
[relayer/src/chain/cosmos.rs:67] "Todo: implement proof verification." = "Todo: implement proof verification."
connection query error Could not parse/unmarshall response: failed to decode Protobuf message: ConnectionEnd.state: invalid wire type: LengthDelimited (expected Varint)
```
The failure is here:
https://github.com/informalsystems/ibc-rs/blob/6f59e2f047777e881fdff79cb8181ed30f76c733/relayer/src/chain/cosmos.rs#L72

The query works on master (with ibc-proto v.0.1.0), the protobuf for `ConnectionEnd` hasn't changed between the two versions.

The channel query that gets a `ChannelEnd` with an almost identical proto for `Sate` works fine.

File: modules/src/ics03_connection/msgs.rs
Comment (ancazamfir) [reply modules/src/ics03_connection/msgs.rs:744]: I think it's fixed now but no idea. Just ran `cargo fmt`, didn't notice it.

File: modules/src/context_mock.rs
Comment (adizere) [inline modules/src/context_mock.rs:46]: Is this buggy or problematic for compilation? Or how come it is commented out?

File: modules/src/ics03_connection/connection.rs
Comment (adizere) [inline modules/src/ics03_connection/connection.rs:25]: I looked for where is this conversion used and could not find. Can you point me to where do we use the decoding of RawIdentifiedConnection -> ConnectionEnd ? Is it the call to `query` from `CosmosSDKChain`?

File: modules/src/ics03_connection/msgs.rs
Comment (adizere) [inline modules/src/ics03_connection/msgs.rs:154]: Did you reorder these functions? I'm just trying to understand why is this diff so complex, because nothing seems to have changed except for the order of functions (and there is `client_state` getter new).

File: modules/src/mock_client/state.rs
Comment (adizere) [inline modules/src/mock_client/state.rs:94]: Let's purge this (or move it in the correct spot, which is probably `mock_client/client_def.rs`).

File: modules/src/ics03_connection/connection.rs
Comment (ancazamfir) [reply modules/src/ics03_connection/connection.rs]: yes it was from the query to fix the connection query with Gaia/ Go Relayer. But as it turns out I was using a slightly older version and once I updated the connection query worked fine. I just reverted this change.

File: modules/src/ics03_connection/msgs.rs
Comment (ancazamfir) [reply modules/src/ics03_connection/msgs.rs:154]: Yes, I ordered the getters consistent with the order in the struct.

File: modules/src/context_mock.rs
Comment (ancazamfir) [reply modules/src/context_mock.rs:44]: Just fixed it.

File: modules/src/ics03_connection/msgs.rs
Comment (adizere) [reply modules/src/ics03_connection/msgs.rs:154]: Good idea, I like this. I'll try to do the same in the future.

**Review by adizere — APPROVED** [review]

**Pull request merged**
